corfudb-tools/src/main/java/org/corfudb/browser/CorfuOfflineBrowserEditor.java

```diff
@@ -0,0 +1,471 @@
+import org.corfudb.protocols.logprotocol.MultiObjectSMREntry;
+import org.corfudb.protocols.logprotocol.MultiSMREntry;
+import org.corfudb.protocols.logprotocol.SMREntry;
+import org.corfudb.protocols.wireprotocol.DataType;
+import org.corfudb.protocols.wireprotocol.IMetadata;
+import org.corfudb.protocols.wireprotocol.LogData;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.CorfuStoreMetadata;
+import org.corfudb.runtime.collections.CorfuTable;
+import org.corfudb.runtime.collections.CorfuDynamicKey;
+import org.corfudb.runtime.collections.CorfuDynamicRecord;
+import org.corfudb.runtime.view.TableRegistry;
+import org.corfudb.util.serializer.DynamicProtobufSerializer;
+
+import javax.annotation.Nonnull;
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.channels.FileChannel;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+import java.util.UUID;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.EnumMap;
+import java.util.Set;
+import java.util.Arrays;
+
+import static org.corfudb.infrastructure.log.StreamLogFiles.parseHeader;
+import static org.corfudb.infrastructure.log.StreamLogFiles.parseMetadata;
+import static org.corfudb.infrastructure.log.StreamLogFiles.parseEntry;
+import static org.corfudb.infrastructure.log.StreamLogFiles.getLogData;
+
+import static org.corfudb.browser.CorfuStoreBrowserEditor.printTableRegistry;
+import static org.corfudb.browser.CorfuStoreBrowserEditor.printKey;
+import static org.corfudb.browser.CorfuStoreBrowserEditor.printMetadata;
+import static org.corfudb.browser.CorfuStoreBrowserEditor.printPayload;
+
+@SuppressWarnings("checkstyle:printLine")
+public class CorfuOfflineBrowserEditor implements CorfuBrowserEditorCommands {
+    private final Path logDir;
+    private DynamicProtobufSerializer dynamicProtobufSerializer;
+
+    private final CorfuTableDescriptor registryTableDsc;
+
+    private final CorfuTableDescriptor protobufDescTableDsc;
+
+    /**
+     * Creates a CorfuOfflineBrowser linked to an existing log directory.
+     * @param offlineDbDir Path to the database directory.
+     */
+    public CorfuOfflineBrowserEditor(String offlineDbDir) throws UncheckedIOException {
+
+        logDir = Paths.get(offlineDbDir);
+        registryTableDsc = new CorfuTableDescriptor(TableRegistry.CORFU_SYSTEM_NAMESPACE,
+                TableRegistry.REGISTRY_TABLE_NAME);
+        protobufDescTableDsc = new CorfuTableDescriptor(TableRegistry.CORFU_SYSTEM_NAMESPACE,
+                TableRegistry.PROTOBUF_DESCRIPTOR_TABLE_NAME);
+    }
+
+
+    public void addRecordsToConcurrentMap(List<LogEntryOrdering> tableEntries, ConcurrentMap cachedTable) {
+
+        tableEntries.sort(new LogEntryComparator());
```

Comment:
The vast majority of the tableEntries coming to this function will likely already be sorted (because offset order will likely match the sequence order) so just wondering if we can keep track of whether the entries are already in sorted order and only sort if necessary?
This is a completely optional optimization.

Reply:
Used a custom Hashmap wrapper class to perform an update by looking into the address/order.